proxy/http2/Http2ConnectionState.cc

```diff
@@ -180,12 +180,14 @@ rcv_data_frame(Http2ConnectionState &cstate, const Http2Frame &frame)
     if (nbytes + read_len > unpadded_length) {
       read_len -= nbytes + read_len - unpadded_length;
     }
-    nbytes += writer->write(myreader, read_len);
-    myreader->consume(nbytes);
+    int num_written = writer->write(myreader, read_len);
+    nbytes += num_written;
+    myreader->consume(num_written);
+    stream->update_read(num_written);
   }
   myreader->writer()->dealloc_reader(myreader);
 
-  if (frame.header().flags & HTTP2_FLAGS_DATA_END_STREAM) {
```

Comment:
Why are we treating read_vio.ntodo() == 0 the same as getting a data end stream flag?

Reply:
Protecting ourselves from ill-written clients that do not send DATA_END_STREAM.  We have seen such clients in our production data stream.  In that case they are sending Content-length, so we can still send READ_COMPLETE when all of the data is written.